src/abi_arm.cpp

```diff
@@ -0,0 +1,102 @@
+// This file is a part of Julia. License is MIT: http://julialang.org/license
+//
+// The ABI implementation used for 32 bit ARM targets.
+
+#ifndef __ARM_EABI__
+#  error "the Julia ARM ABI implementation only supports EABI"
+#endif
+
+#ifndef __ARM_PCS_VFP
+#  error "the Julia ARM ABI implementation requires VFP support"
+#endif
+
+typedef bool AbiState;
+AbiState default_abi_state = 0;
+
+inline bool isHomogeneousAggregate(jl_value_t *ty, const jl_value_t *&base,
+                                   size_t &members) {
+    if (jl_is_structtype(ty)) {
+        members = jl_datatype_nfields(ty);
+        for (size_t i = 0; i < members; ++i) {
+            jl_value_t *fty = jl_field_type((jl_datatype_t*)ty,i);
+            if (!base)
+                base = fty;
+            else if (base != fty)
+                return false;
+        }
+        return true;
+    }
+
+    return false;
+}
+
+bool use_sret(AbiState *state,jl_value_t *ty)
+{
+    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
+        return false;
+    size_t size = jl_datatype_size(ty);
+
+    // A double-word sized Fundamental Data Type (e.g., long long, double and
+    // 64-bit containerized vectors) is returned in r0 and r1.
+    if (jl_is_bitstype(ty) && size <= 8)
+        return false;
+
+    // A Composite Type not larger than 4 bytes is returned in r0. The format is
+    // as if the result had been stored in memory at a word-aligned address and
+    // then loaded into r0 with an LDR instruction. Any bits in r0 that lie
+    // outside the bounds of the result have unspecified values.
+    if (jl_is_structtype(ty) && size <= 4)
+        return false;
+
+    // For the VFP the following argument types are VFP CPRCs:
+    // - ...
+    // - A Homogeneous Aggregate with a Base Type of a single- or
+    //   double-precision floating-point type with one to four Elements.
+    // Any result whose type would satisfy the conditions for a VFP CPRC is
+    // returned in the appropriate number of consecutive VFP registers
+    const jl_value_t *base = nullptr;
+    size_t members = 0;
+    if (isHomogeneousAggregate(ty, base, members) && members <= 4 &&
+        (base == (jl_value_t*)jl_float64_type ||
+         base == (jl_value_t*)jl_float32_type))
+        return false;
+
+    return true;
+}
+
+void needPassByRef(AbiState *state,jl_value_t *ty, bool *byRef, bool *inReg)
+{
+    return;
+}
+
+Type *preferred_llvm_type(jl_value_t *ty, bool isret)
+{
+    // TODO: If the argument is an integral Fundamental Data Type that is
+    // smaller than a word, then it is zero- or sign-extended to a full word and
+    // its size is set to 4 bytes. If the argument is a Half-precision Floating
+    // Point Type it is converted to Single Precision.
```

Comment:
you can return `T_int32` and the callee of this function will zero/sign extend to accommodate